openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+/// levelset. This is done by means of a sparse and parallel fast sweeping
+/// algorithm based on a first order Goudonov's scheme.
+///
+/// Solves: @f$|\nabla \phi|^2 = 1 @f$
+///
+/// @details The grid needs to have correct values set near the zero level set, where "near"
+/// means 1.5 times the size of a grid voxel.  These near values are not modified.
+/// Use multiple iterations (@a nIter > 1) for complex geometry where characteristics cross
+/// (along the medial axis). Each iteration performs eight sweeps corresponding to
+/// the grouping of characteristics in the eight coordinate quadrants.
+///
+/// @param grid   a grid containing initial zero-crossing signed distance values
+///               (@f$|\phi|@f$ < 1.5*voxel size).
+/// @param nIter  the number of interations to be performed (each with 8 sweeps).
+///
+/// @note Multi-threaded and sparse!
+///
+/// @warning This method only works for scalar, floating-point grids
+/// and it is @e not the preferred way to solve the eikonal equation for
+/// narrow-band level sets (of width ~2x3). Rather it is an efficient
+/// technique when distance values must be computed far from known
+/// distance values (typically close to an interface).
+///
+/// @warning This class is for experts only (the sequence in with the different
+///          methods are called is important). Instead call one of the free-standing
+///          functions listed above!
+template<typename GridT>
+class FastSweeping
+{
+    using ValueT = typename GridT::ValueType;
+    using TreeT  = typename GridT::TreeType;
+    using AccT   = tree::ValueAccessor<TreeT, false>;//don't register accessors
+
+    // This class can only be templated on a grid with floating-point values!
+    static_assert(std::is_floating_point<ValueT>::value,
+                  "FastSweeping requires a grid with floating-point values");
+public:
+
+    /// @brief Constructor
+    FastSweeping();
+
+     /// @brief Destructor.
+    ~FastSweeping() { this->clear(); }
+
+    /// @brief Disallow copy construction.
+    FastSweeping(const FastSweeping&) = delete;
+
+    /// @brief Disallow copy assignment.
+    FastSweeping& operator=(const FastSweeping&) = delete;
+
+    /// @brief Returns a shared pointer to the signed distance field computed
+    ///        by this class.
+    /// @warning This shared pointer might point to NULL if the grid has not been
+    ///          initialize (by one of the init methods) or computed (by the sweep
+    ///          method).
+    typename GridT::Ptr sdfGrid() { return mGrid1; }
+    typename GridT::Ptr extGrid() { return mGrid2; }
+
+    void initSdf(const GridT &fogGrid, ValueT isoValue, bool isInputSdf);
+
+    template <typename OpT>
+    void initExt(const GridT &fogGrid, const OpT &op, ValueT isoValue, bool isInputSdf);
+
+    // use tools::NN_FACE_EDGE for improved dilation
+    void initDilate(const GridT &sdfGrid, int dilation, NearestNeighbors nn = NN_FACE);
```

Comment:
What does the above comment refer to, is this a todo? What's the advantage over performing face edge dilation instead of face or face edge vertex in context of these algorithms

Reply:
just that - dilation with higher connectivity is obviously slower but arguable also more accurate (depending on the application ). I can remove this comment if it bothers you